contract/migrations/18.0.2.0.0/pre-migrate.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2025 ACSONE SA/NV
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    env.cr.execute(
+        """
+    DELETE FROM ir_model
+    WHERE model IN
+        ('contract.abstract.contract',
+        'contract.abstract.contract.line',
+        'contract.recurrency.mixin',
+        'contract.recurrency.basic.mixin'
+    );
+    """
+    )
```

Comment:
```suggestion
    )
    views_to_delete = ["contract.contract_template_line_form_view"]
    for view_ref in views_to_delete:
        view = env.ref(view_ref, raise_if_not_found=False)
        if view:
            view.unlink()
```

I got an error during the upgrade that **is_auto_renew** is not part of the module. This occures duing the update of **contract_line_form_view** which references **contract_abstract_contract_line_form_view**.

Reply:
The problem with this migration script is that you cannot delete this view if other views inherit from it.

Reply:
But you shouldn't remove views in the middle of the migration. OpenUpgrade already avoids a hard crash on that case, and they should be removed at the end of the cycle being noupdate=0. If they are not remvoed, it's because there are manual children views, and in that case, you should fix it manually on your DB.

Reply:
Is there a possibility to force the update of **contract_template_line_form_view** before **contract_line_form_view**  is updated? This could also fix the issue without removing/readding the contract_template_line_form_view.

Reply:
I insist: which issue? How are you "updating"?

Reply:
@pedrobaeza I got the same issue when launching a `click-odoo-update --update-all` on a DB that was migrated by Odoo from v12 to v18.
@CRogos I personally solved my issue by deleting all views (all entries in ir_ui_view table) before launching this big click-odoo-update, as in our case the users are not supposed to modify the views from the UI.

Reply:
FYI [delete_records_safely_by_xml_id](https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L3428 ) contains `delete_childs` parameter to delete all inheriting views of the view you want to delete.

Reply:
@MiquelRForgeFlow solution is the best in my opinion, because everything which refers to this view should be recreated later anyway. (Anyone who is allowing to modify views in the UI can't be helped anyway 🤪 )

Reply:
Have not tested yet:

```suggestion
    )
    views_to_delete = ["contract.contract_template_line_form_view"]
    openupgrade.delete_records_safely_by_xml_id(env, views_to_delete, True) 
```

Reply:
@sbejaoui could you add the above suggestion to the migration script. 

I still get this error when upgrading my database with odoo.sh.

```
2025-06-03 08:06:56,339 99 ERROR  odoo.modules.registry: Failed to load registry
2025-06-03 08:06:56,343 99 CRITICAL  odoo.service.server: Failed to initialize database ` `. 
Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/service/server.py", line 1328, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/tools/func.py", line 97, in locked
    return func(inst, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/registry.py", line 127, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 480, in load_modules
    processed_modules += load_marked_modules(env, graph,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 365, in load_marked_modules
    loaded, processed = load_module_graph(
                        ^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 228, in load_module_graph
    load_data(env, idref, mode, kind='data', package=package)
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 72, in load_data
    tools.convert_file(env, package.name, filename, idref, mode, noupdate, kind)
  File "/home/odoo/src/odoo/odoo/tools/convert.py", line 608, in convert_file
    convert_xml_import(env, module, fp, idref, mode, noupdate)
  File "/home/odoo/src/odoo/odoo/tools/convert.py", line 679, in convert_xml_import
    obj.parse(doc.getroot())
  File "/home/odoo/src/odoo/odoo/tools/convert.py", line 594, in parse
    self._tag_root(de)
  File "/home/odoo/src/odoo/odoo/tools/convert.py", line 548, in _tag_root
    raise ParseError(msg) from None  # Restart with "--log-handler odoo.tools.convert:DEBUG" for complete traceback
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
odoo.tools.convert.ParseError: while parsing /home/odoo/src/user/modules/oca/contract/contract/views/contract_template_line.xml:4
Error while validating view near:

                    </group>
                    <group invisible="display_type">
                        <field name="is_auto_renew"/>
                        <field name="is_canceled" invisible="1"/>
                    </group>

Field "is_auto_renew" does not exist in model "contract.template.line"

View error context:
{'file': '/home/odoo/src/user/modules/oca/contract/contract/views/contract_template_line.xml',
 'line': 33,
 'name': 'contract.template.line form view (in contract)',
 'view': ir.ui.view(3100,),
 'view.model': 'contract.template.line',
 'view.parent': ir.ui.view(3086,),
 'xmlid': 'contract_template_line_form_view'}
```